File: modules/vm-webrtc/src/VmWebrtcModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

import { log } from '../../../lib/logger';
import {
  createGithubConnectorTool,
  type GithubConnectorNativeModule,
} from './ToolGithubConnector';
import { gdriveConnectorDefinition } from './ToolGDriveConnector';
import {
  createGPT5GDriveFixerTool,
  type GPT5GDriveFixerNativeModule,
} from './ToolGPT5GDriveFixer';
import {
  createGPT5WebSearchTool,
  type GPT5WebSearchNativeModule,
} from './ToolGPT5WebSearch';
import {
  createToolkitHelper,
  type ToolkitHelperNativeModule,
} from './ToolkitHelper';
import type { VadMode } from '../../../lib/vadPreference';
import toolManager from './ToolManager';
import {
    OpenAIConnectionOptions,
    OpenAIConnectionState,
    VmWebrtcModuleEvents,
} from './VmWebrtc.types';
import { getToolkitDefinitions } from './ToolkitManager';

const MODULE_NAME = 'VmWebrtc';

const makeUnavailableError = () =>
  new Error(`Native module ${MODULE_NAME} is unavailable. Rebuild the iOS app to load native code.`);

const loadModule = () => requireOptionalNativeModule<VmWebrtcModule>(MODULE_NAME);

declare class VmWebrtcModule extends NativeModule<VmWebrtcModuleEvents> {
  PI: number;
  hello(): string;
  helloFromExpoModule(): string;
  setValueAsync(value: string): Promise<void>;
  openOpenAIConnectionAsync(options: OpenAIConnectionOptions): Promise<OpenAIConnectionState>;
  closeOpenAIConnectionAsync(): Promise<OpenAIConnectionState>;
  githubOperationFromSwift(query: string): Promise<string>;
  sendGithubConnectorResponse(requestId: string, result: string): void;
  gpt5GDriveFixerOperationFromSwift(paramsJson: string): Promise<string>;
  sendGPT5GDriveFixerResponse(requestId: string, result: string): void;
  gpt5WebSearchOperationFromSwift(query: string): Promise<string>;
  sendGPT5WebSearchResponse(requestId: string, result: string): void;
  sendToolkitResponse(requestId: string, result: string): void;
  muteUnmuteOutgoingAudio(shouldMute: boolean): void;
  initializeLogfireTracing(serviceName: string, apiKey: string): Promise<void>;
  logfireEvent(tracerName: string, spanName: string, attributes?: Record<string, unknown>): void;
}

const module = loadModule();

if (!module) {
  log.warn(`[${MODULE_NAME}] Native module not found. Did you rebuild the iOS app?`);
}

// Initialize github connector tool
const githubConnectorTool = createGithubConnectorTool(module as unknown as GithubConnectorNativeModule | null);
const gpt5GDriveFixerTool = createGPT5GDriveFixerTool(
  module as unknown as GPT5GDriveFixerNativeModule | null,
  gdriveConnectorDefinition.description,
);
const gpt5WebSearchTool = createGPT5WebSearchTool(
  module as unknown as GPT5WebSearchNativeModule | null,
);

// Initialize Gen2 toolkit helper
const toolkitHelper = createToolkitHelper(module as unknown as ToolkitHelperNativeModule | null);

export const helloFromExpoModule = () => {
  if (!module) {
    throw makeUnavailableError();
  }

  return module.helloFromExpoModule();
};

export const openOpenAIConnectionAsync = async (
  options: OpenAIConnectionOptions
) => {
  if (!module) {
    throw makeUnavailableError();
  }

  const trimmedVoice = options.voice?.trim();
  const resolvedVoice = trimmedVoice && trimmedVoice.length > 0 ? trimmedVoice : 'cedar';
  const trimmedInstructions = options.instructions.trim();
  if (trimmedInstructions.length === 0) {
    throw new Error(`[${MODULE_NAME}] instructions must be a non-empty string.`);
  }
  // const toolDefinitionsWithPrompts = await toolManager.getAugmentedToolDefinitions(
  //   options.toolDefinitions,
  // );

  // log.info(`[${MODULE_NAME}] Tool definitions resolved`, {}, {
  //   definitions: toolDefinitionsWithPrompts,
  // });

  // Get Gen2 toolkit definitions already converted to ToolDefinition format with qualified names
  const toolDefinitionsFromToolkits = getToolkitDefinitions(); // gen2
  log.info(`[${MODULE_NAME}] Toolkit definitions resolved`, {}, {
    definitions: toolDefinitionsFromToolkits,
  });

  // Merge Gen1 and Gen2 tool definitions
  // const mergedToolDefinitions = [
  //   ...(toolDefinitionsWithPrompts || []),
  //   ...toolDefinitionsFromToolkits,
  // ];

  const mergedToolDefinitions = toolDefinitionsFromToolkits;

  log.info(`[${MODULE_NAME}] Merged tool definitions`, {}, {
    definitions: mergedToolDefinitions,
  });

  const resolvedVadMode: VadMode = options.vadMode === 'semantic' ? 'semantic' : 'server';

  const resolvedAudioSpeed =
    typeof options.audioSpeed === 'number'
      ? Math.min(Math.max(options.audioSpeed, 0.25), 4)
      : undefined;

  const sanitizedOptions: OpenAIConnectionOptions = {
    ...options,
    voice: resolvedVoice,
    instructions: trimmedInstructions,
    toolDefinitions: mergedToolDefinitions,
    vadMode: resolvedVadMode,
    audioSpeed: resolvedAudioSpeed,
  };

  log.debug(`[${MODULE_NAME}] openOpenAIConnectionAsync invoked`, {}, {
    hasBaseUrl: Boolean(options.baseUrl),
    hasModel: Boolean(options.model),
    audioOutput: options.audioOutput ?? 'handset',
    audioSpeed: resolvedAudioSpeed ?? 'default',
    hasInstructions: trimmedInstructions.length > 0,
    voice: resolvedVoice,
    vadMode: resolvedVadMode,
  });

  return module.openOpenAIConnectionAsync(sanitizedOptions);
};

export const closeOpenAIConnectionAsync = () => {
  if (!module) {
    throw makeUnavailableError();
  }

  log.debug(`[${MODULE_NAME}] closeOpenAIConnectionAsync invoked`);

  return module.closeOpenAIConnectionAsync();
};

export const muteUnmuteOutgoingAudio = (shouldMute: boolean) => {
  if (!module) {
    throw makeUnavailableError();
  }

  log.debug(`[${MODULE_NAME}] muteUnmuteOutgoingAudio invoked`, {}, { shouldMute });
  module.muteUnmuteOutgoingAudio(shouldMute);
};

// Github Connector function that can be called from Swift
// Uses the Github API
export const githubConnector = async (params: { self_contained_javascript_octokit_code_snippet: string }): Promise<string> => {
  if (!githubConnectorTool) {
    throw makeUnavailableError();
  }

  return githubConnectorTool.execute(params);
};

// Github Connector bridge function - calls JS github connector from Swift
export const githubOperationFromSwift = async (codeSnippet: string): Promise<string> => {
  if (!githubConnectorTool) {
    throw makeUnavailableError();
  }

  return githubConnectorTool.executeFromSwift(codeSnippet);
};

// Export tool instances for direct access if needed
export { githubConnectorTool, gpt5GDriveFixerTool, gpt5WebSearchTool };

// This call loads the native module object from the JSI.
export default (module ?? ({} as VmWebrtcModule));

File: modules/vm-webrtc/ios/VmWebrtcModule.swift
import ExpoModulesCore
import Foundation

struct OpenAIConnectionOptions: Record {
  @Field
  var apiKey: String

  @Field
  var model: String?

  @Field
  var baseUrl: String?

  @Field
  var audioOutput: String?

  @Field
  var instructions: String

  @Field
  var voice: String?

  @Field
  var toolDefinitions: [[String: Any]]?

  @Field
  var vadMode: String?

  @Field
  var audioSpeed: Double?

  @Field
  var enableRecording: Bool?

  @Field
  var maxConversationTurns: Int?

  @Field
  var retentionRatio: Double?
}

public class VmWebrtcModule: Module {
  private lazy var webrtcClient = OpenAIWebRTCClient()
  private var toolGithubConnector: ToolGithubConnector?
  // Add GDrive connector tool instance
  private var toolGDriveConnector: ToolGDriveConnector?
  private var toolGPT5GDriveFixer: ToolGPT5GDriveFixer?
  private var toolGPT5WebSearch: ToolGPT5WebSearch?
  // Add Gen2 toolkit helper
  private var toolkitHelper: ToolkitHelper?
  private let logfireTracingManager = LogfireTracingManager()
  private var logger: NativeLogger { VmWebrtcLogging.logger }

  public func helloFromExpoModule() -> String {
    return "Hello world from module"
  }

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('VmWebrtc')` in JavaScript.
    Name("VmWebrtc")

    // Defines constant property on the module.
    Constant("PI") {
      Double.pi
    }

    // Defines event names that the module can send to JavaScript.
    Events(
      "onChange",
      "onGithubConnectorRequest",
      "onGithubConnectorResponse",
      // Add GDrive events
      "onGDriveConnectorRequest",
      "onGDriveConnectorResponse",
      "onGPT5GDriveFixerRequest",
      "onGPT5GDriveFixerResponse",
      "onGPT5WebSearchRequest",
      "onGPT5WebSearchResponse",
      // Add Gen2 toolkit events
      "onToolkitRequest",
      "onToolkitResponse",
      "onIdleTimeout",
      "onTokenUsage",
      "onRealtimeError",
      "onAudioMetrics"
    )

    // Initialize native tool delegates used by the module
    OnCreate {
      VmWebrtcLogging.configureTracingManager(self.logfireTracingManager)
      self.logger.log("OnCreate: initializing tool delegates")
      self.webrtcClient.setEventEmitter { [weak self] eventName, payload in
        guard let self else { return }
        self.sendEvent(eventName, payload)
      }
      self.logger.log("Event emitter configured for OpenAI WebRTC client")
      // Initialize github connector tool
      self.toolGithubConnector = ToolGithubConnector(module: self, responder: self.webrtcClient)
      let githubInitialized = self.toolGithubConnector != nil
      self.logger.log("ToolGithubConnector initialized = \(githubInitialized)")

      // Initialize gdrive connector tool
      self.toolGDriveConnector = ToolGDriveConnector(module: self, responder: self.webrtcClient)
      let gdriveInitialized = self.toolGDriveConnector != nil
      self.logger.log("ToolGDriveConnector initialized = \(gdriveInitialized)")

      // Initialize GPT5 gdrive fixer tool
      self.toolGPT5GDriveFixer = ToolGPT5GDriveFixer(module: self, responder: self.webrtcClient)
      let fixerInitialized = self.toolGPT5GDriveFixer != nil
      self.logger.log("ToolGPT5GDriveFixer initialized = \(fixerInitialized)")

      // Initialize GPT5 web search tool
      self.toolGPT5WebSearch = ToolGPT5WebSearch(module: self, responder: self.webrtcClient)
      let webSearchInitialized = self.toolGPT5WebSearch != nil
      self.logger.log("ToolGPT5WebSearch initialized = \(webSearchInitialized)")

      // Initialize Gen2 toolkit helper
      self.toolkitHelper = ToolkitHelper(module: self, responder: self.webrtcClient)
      let toolkitInitialized = self.toolkitHelper != nil
      self.logger.log("ToolkitHelper initialized = \(toolkitInitialized)")

      // Wire delegates into the WebRTC client
      self.webrtcClient.setGithubConnectorDelegate(self.toolGithubConnector!)
      if let gdrive = self.toolGDriveConnector {
        self.webrtcClient.setGDriveConnectorDelegate(gdrive)
        self.logger.log("Delegate set: gdrive")
      } else {
        self.logger.log("Delegate set NOT: gdrive")
      }
      if let fixer = self.toolGPT5GDriveFixer {
        self.webrtcClient.setGPT5GDriveFixerDelegate(fixer)
        self.logger.log("Delegate set: GPT5 fixer")
      } else {
        self.logger.log("Delegate set NOT: GPT5 fixer")
      }
      if let webSearch = self.toolGPT5WebSearch {
        self.webrtcClient.setGPT5WebSearchDelegate(webSearch)
        self.logger.log("Delegate set: GPT5 web search")
      } else {
        self.logger.log("Delegate set NOT: GPT5 web search")
      }
      if let toolkit = self.toolkitHelper {
        self.webrtcClient.setToolkitHelper(toolkit)
        self.logger.log("Delegate set: Toolkit helper")
      } else {
        self.logger.log("Delegate set NOT: Toolkit helper")
      }
      self.logger.log("Delegates set: github, possibly GDrive, toolkit")
    }

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      return "Hello world! 👋"
    }

    Function("helloFromExpoModule") { () -> String in
      return self.helloFromExpoModule()
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("setValueAsync") { (value: String) in
      // Send an event to JavaScript.
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    AsyncFunction("openOpenAIConnectionAsync") { (options: OpenAIConnectionOptions) -> String in
      self.logger.log("openOpenAIConnectionAsync called", attributes: [
        "model": options.model ?? "nil",
        "audioOutput": options.audioOutput ?? "nil",
        "voice": options.voice ?? "nil",
        "vadMode": options.vadMode ?? "nil"
      ])
      let outputPreference = OpenAIWebRTCClient.AudioOutputPreference(
        rawValue: options.audioOutput ?? "handset"
      ) ?? .handset

      let sanitizedInstructions = options.instructions
        .trimmingCharacters(in: .whitespacesAndNewlines)

      guard !sanitizedInstructions.isEmpty else {
        throw NSError(
          domain: "VmWebrtc",
          code: 1001,
          userInfo: [
            NSLocalizedDescriptionKey: "instructions must be a non-empty string."
          ]
        )
      }

      let toolDefinitions = options.toolDefinitions ?? []
      await MainActor.run {
        self.webrtcClient.setToolDefinitions(toolDefinitions)
      }

      let state = try await self.webrtcClient.openConnection(
        apiKey: options.apiKey,
        model: options.model,
        baseURL: options.baseUrl,
        audioOutput: outputPreference,
        instructions: sanitizedInstructions,
        voice: options.voice,
        vadMode: options.vadMode,
        audioSpeed: options.audioSpeed,
        enableRecording: options.enableRecording ?? false,
        maxConversationTurns: options.maxConversationTurns,
        retentionRatio: options.retentionRatio
      )
      return state
    }

    AsyncFunction("closeOpenAIConnectionAsync") { () -> String in
      let state = await MainActor.run {
        self.webrtcClient.closeConnection()
      }
      return state
    }

    AsyncFunction("initializeLogfireTracing") { (serviceName: String, apiKey: String) in
      try await self.logfireTracingManager.initialize(serviceName: serviceName, apiKey: apiKey)
    }

    Function("logfireEvent") { (tracerName: String, spanName: String, attributes: [String: Any]?) in
      let resolvedSeverity = LogfireTracingManager.severity(from: attributes) ?? .info
      print(
        "[VmWebrtcModule] logfireEvent span=\(spanName) severityText=\(resolvedSeverity.severityText) severityNumber=\(resolvedSeverity.severityNumber)"
      )
      self.logfireTracingManager.recordEvent(
        tracerName: tracerName,
        spanName: spanName,
        attributes: attributes,
        severity: resolvedSeverity,
        severityText: resolvedSeverity.severityText,
        severityNumber: resolvedSeverity.severityNumber
      )
    }

    // JavaScript calls this to send github connector result back
    Function("sendGithubConnectorResponse") { (requestId: String, result: String) in
      self.logger.log("JS→Native sendGithubConnectorResponse", attributes: [
        "requestId": requestId,
        "result_length": result.count
      ])
      self.toolGithubConnector?.handleResponse(requestId: requestId, result: result)
    }

    // Add: JavaScript calls this to send GDrive connector result back
    Function("sendGDriveConnectorResponse") { (requestId: String, result: String) in
      self.logger.log("JS→Native sendGDriveConnectorResponse", attributes: [
        "requestId": requestId,
        "result_length": result.count,
        "result_preview": String(result.prefix(1000))
      ])
      self.toolGDriveConnector?.handleResponse(requestId: requestId, result: result)
    }

    Function("sendGPT5GDriveFixerResponse") { (requestId: String, result: String) in
      self.logger.log("JS→Native sendGPT5GDriveFixerResponse", attributes: [
        "requestId": requestId,
        "result_length": result.count
      ])
      self.toolGPT5GDriveFixer?.handleResponse(requestId: requestId, result: result)
    }

    Function("sendGPT5WebSearchResponse") { (requestId: String, result: String) in
      self.logger.log("JS→Native sendGPT5WebSearchResponse", attributes: [
        "requestId": requestId,
        "result_length": result.count
      ])
      self.toolGPT5WebSearch?.handleResponse(requestId: requestId, result: result)
    }

    // Github Connector function - calls JavaScript github connector via events
    AsyncFunction("githubOperationFromSwift") { (codeSnippet: String, promise: Promise) in
      self.logger.log("Swift→JS githubOperationFromSwift", attributes: [
        "snippet_length": codeSnippet.count
      ])
      self.toolGithubConnector?.githubOperationFromSwift(codeSnippet: codeSnippet, promise: promise)
    }

    // GDrive bridge: call JS GDrive connector via events for Swift testing
    AsyncFunction("gdriveOperationFromSwift") { (codeSnippet: String, promise: Promise) in
      self.toolGDriveConnector?.gdriveOperationFromSwift(codeSnippet: codeSnippet, promise: promise)
    }

    AsyncFunction("gpt5GDriveFixerOperationFromSwift") { (paramsJson: String, promise: Promise) in
      self.toolGPT5GDriveFixer?.gpt5GDriveFixerOperationFromSwift(paramsJson: paramsJson, promise: promise)
    }

    AsyncFunction("gpt5WebSearchOperationFromSwift") { (query: String, promise: Promise) in
      self.toolGPT5WebSearch?.gpt5WebSearchOperationFromSwift(query: query, promise: promise)
    }

    Function("muteUnmuteOutgoingAudio") { (shouldMute: Bool) in
      Task { @MainActor in
        self.webrtcClient.setOutgoingAudioMuted(shouldMute)
      }
    }

  }
}
